### PR TITLE
[PATCH v4] Build and test ODP with shared DPDK

### PR DESCRIPTION
**Pull request by lumag**
Another PR dedicated to distro support. Build and test ODP with shared version of DPDK.

### CODE REVIEW

**Comment by Bill-Fischofer-Linaro** [comment]
@lumag Looks like recent merges now require this PR to be rebased.

**Comment by lumag** [comment]
@Bill-Fischofer-Linaro done

File: .travis.yml [hunk 1]
```diff
@@ -198,7 +199,14 @@ install:
...
         - |
           if [ -n "$TARGET" ] ; then
-           if [ ! -f "dpdk/${TARGET}/lib/libdpdk.a" ] ; then
+           # force DPDK rebuild if it does not match shared/non-shared requirement
+           if [ -n "$DPDK_SHARED" -a ! -f "dpdk/${TARGET}/lib/libdpdk.so" ] ; then
+            rm -rf dpdk/"$TARGET"
+           fi
+           if [ -z "$DPDK_SHARED" -a ! -f "dpdk/${TARGET}/lib/libdpdk.a" ] ; then
+            rm -rf dpdk/"$TARGET"
+           fi
```
Comment (muvarov) [inline .travis.yml]: maybe simple  
`if [  -a ! -f "dpdk/${TARGET}/lib/libdpdk.a" ] ; then`
instead of 2 checks?

Comment (muvarov) [reply .travis.yml]: In general that is not right solution. I think we need to cache 2 dpdk builds. Shared and non shared and force dpdk recompilation only on version change.

File: .travis.yml [hunk 2]
```diff
@@ -258,11 +271,17 @@ script:
...
         - if [ -z "$CROSS_ARCH" ] ; then
-          LD_LIBRARY_PATH="${HOME}/odp-install/lib:$LD_LIBRARY_PATH" ./odp_hello_inst_dynamic ;
-          ./odp_hello_inst_static ;
+            LD_LIBRARY_PATH="${HOME}/odp-install/lib:$LD_LIBRARY_PATH" ./odp_hello_inst_dynamic ;
+          fi
+        - echo "Static link.."
+        - |
+          # it is not possible to do static linking if we only have shared DPDK library. Compiler complains about missing -ldpdk
+          if [ -z "$TARGET" -o -z "$DPDK_SHARED" ] ; then
+            ${CC} ${CFLAGS} ${OLDPWD}/example/hello/odp_hello.c -o odp_hello_inst_static `PKG_CONFIG_PATH=${HOME}/odp-install/lib/pkgconfig:${PKG_CONFIG_PATH} pkg-config --cflags --libs libodp-linux --static` -static || exit 1
```
Comment (muvarov) [inline .travis.yml]: why exit 1 is needed? Travis should stop on error return code on the command. exit 1 should not be needed.

File: .travis.yml [hunk 3]
```diff
@@ -258,11 +271,17 @@ script:
...
+            if [ -z "$CROSS_ARCH" ] ; then
+            ./odp_hello_inst_static || exit 1;
```
Comment (muvarov) [inline .travis.yml]: same here.

**Comment by muvarov** [comment]
"build: another fix for linking with shared libdpdk"  is better to rename to "build: odp_dpdk.m4: another fix for linking with shared libdpdk". It will be good to add some readable description what was a problem with previous version.

File: .travis.yml (hunk 1 above)
Comment (lumag) [reply .travis.yml]: Hmm. I can try doing that.

File: .travis.yml (hunk 2 above)
Comment (lumag) [reply .travis.yml]: exit 1 is needed, to actually return the error code out of if.

Comment (lumag) [reply .travis.yml]: Hmm. Maybe it's not that necessary.

File: .travis.yml (hunk 3 above)
Comment (lumag) [reply .travis.yml]: Generally first `exit 1` is necessary, otherwise `if` will return the status of last command, not the first failed one.

Comment (lumag) [reply .travis.yml]: Try `if true ; then false; if false ; then true ; fi; fi; echo $?`

**Comment by lumag** [comment]
@muvarov updated PR to follow your suggestions.

**Comment by muvarov** [comment]
Merged.

**Pull request closed without merging**